lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,679 @@
+    #[error("Unable to load the bindings for \"{module}\" because \"{}\" doesn't exist", path.display())]
+    MissingBindings { module: String, path: PathBuf },
+    #[error("Error building package when parsing module \"{0}\": {1}.")]
+    ErrorBuildingPackage(String, io::Error),
+    #[error(
+        "Path \"{0}\", specified in the manifest as part of the package file system does not exist.",
+    )]
+    MissingManifestFsPath(String),
+    #[error("When processing the package filesystem, found path \"{0}\" which is not a directory")]
+    PackageFileSystemEntryMustBeDirectory(String),
+}
+
+impl Publish {
+    /// Executes `wasmer publish`
+    pub fn execute(&self) -> Result<(), anyhow::Error> {
+        let mut builder = Builder::new(Vec::new());
+
+        let cwd = match self.package_path.as_ref() {
+            Some(s) => std::env::current_dir()?.join(s),
+            None => std::env::current_dir()?,
+        };
+
+        let manifest_path_buf = cwd.join("wasmer.toml");
+        let manifest = std::fs::read_to_string(&manifest_path_buf)
+            .map_err(|e| anyhow::anyhow!("could not find manifest: {e}"))
+            .with_context(|| anyhow::anyhow!("{}", manifest_path_buf.display()))?;
+        let mut manifest = wapm_toml::Manifest::parse(&manifest)?;
+        manifest.base_directory_path = cwd.clone();
+
+        if let Some(package_name) = self.package_name.as_ref() {
+            manifest.package.name = package_name.to_string();
+        }
+
+        if let Some(version) = self.version.as_ref() {
+            manifest.package.version = version.clone();
+        }
+
+        // See if a user is logged in. The backend should check for authorization on uploading
+        let (registry, username) =
+            wasmer_registry::whoami(self.registry.as_deref(), self.token.as_deref()).with_context(
+                || {
+                    anyhow::anyhow!(
+                        "could not find username / registry for registry = {:?}, token = {}",
+                        self.registry,
+                        self.token.as_deref().unwrap_or_default()
+                    )
+                },
+            )?;
+
+        let registry_present =
+            wasmer_registry::test_if_registry_present(&registry).unwrap_or(false);
+
+        if !registry_present {
+            return Err(anyhow::anyhow!(
+                "registry {} is currently unavailable",
+                registry
+            ));
+        }
+
+        // If the package name is not a "/", try to prefix it with the current logged in user
+        if !Package::validate_package_name(&manifest.package.name) {
+            manifest.package.name = format!("{username}/{}", manifest.package.name);
+        }
+
+        // Validate the name again
+        if !Package::validate_package_name(&manifest.package.name) {
+            return Err(anyhow::anyhow!(
+                "Invalid package name {:?}",
+                manifest.package.name
+            ));
+        }
```

Comment:
We chatted about this on slack, we mentioned to not do any validation on the package name on the clients whatsoever (just for wasmer run as an optimization mechanism).

Reply:
Done in https://github.com/wasmerio/wasmer/pull/3432/commits/46a2d28f30de284ddef3bff417c0faa6b7e627dc